crates/ide/src/ide/assists/add_to_lambda.rs

```diff
@@ -0,0 +1,84 @@
+//! add an undefined name to the top-level lambda
+use super::{AssistKind, AssistsCtx};
+use crate::def::Name;
+use crate::TextEdit;
+use syntax::ast::{self, AstNode};
+use syntax::{TextRange, TextSize};
+
+pub(super) fn add_to_lambda(ctx: &mut AssistsCtx<'_>) -> Option<()> {
+    let name = ctx.covering_node::<ast::Ref>()?.token()?;
+    let name = name.text();
+
+    // the whole file should be a lambda
+    let pat = ast::Lambda::cast(ctx.ast.syntax().first_child()?)?
+        .param()?
+        .pat()?;
+
+    // name should be undefined
+    if ctx
+        .db
+        .module(ctx.frange.file_id)
+        .names()
+        .any(|(_, Name { text, .. })| text == name)
+    {
+        return None;
+    };
+
+    let (pos, insert) = if let Some(field) = pat.fields().last() {
+        let pos = field.syntax().text_range().end();
+        let left = pos - TextSize::from(1);
+        // insert before the space if the field ends with a space
+        (
+            if ctx.ast.syntax().text().char_at(left) == Some(' ') {
+                left
+            } else {
+                pos
+            },
+            format!(", {name}"),
+        )
```

Comment:
```suggestion
        let mut pos = field.syntax().text_range().end();
        // Insert before the space if the PatField ends with a space.
        if field.syntax().text().ends_with(' ') {
            pos -= 1;
        }
        (pos, format!(", {name}"))
```

Still not a fan of this patch work. Multiple spaces in source could produce weird results. But I think it's mostly fine.

Reply:
This didn't work for me, I don't think `SyntaxText` has `ends_with`

Reply:
> This didn't work for me, I don't think SyntaxText has ends_with

Oh I forget it's the `SyntaxText`. Maybe `matches!(field.syntax().last_token(), Some(tok) if tok.text().ends_with(' '))`
Whitespace handling is quite annoying now. 
There's an RFC to improve trivia handling https://github.com/rust-lang/rust-analyzer/issues/6584, but seems stall.